internal/query/roles_v2.go

```diff
@@ -0,0 +1,668 @@
+
+	var (
+		actions []string
+		dbrole  storage.Role
+		err     error
+		errs    = make(chan error, ReadRolesErrBufLen)
+		wg      = &sync.WaitGroup{}
+	)
+
+	ctx, span := e.tracer.Start(
+		ctx,
+		"engine.GetRoleV2",
+		trace.WithAttributes(attribute.Stringer("role", role.ID)),
+	)
+	defer span.End()
+
+	// check if the role is a valid v2 role
+	if role.Type != e.rbac.RoleResource.Name {
+		err := fmt.Errorf("%w: %s is not a valid v2 Role", ErrInvalidType, role.Type)
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.Role{}, err
+	}
+
+	// 1. Get role actions from spice DB
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		spicedbctx, span := e.tracer.Start(ctx, "listRoleV2Actions")
+		defer span.End()
+
+		actions, err = e.listRoleV2Actions(spicedbctx, types.Role{ID: role.ID})
+		if err != nil {
+			errs <- err
+			return
+		}
+	}()
+
+	// 2. Get role info (name, created_by, etc.) from permissions API DB
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		apidbctx, span := e.tracer.Start(ctx, "getRoleFromPermissionAPI")
+		defer span.End()
+
+		dbrole, err = e.store.GetRoleByID(apidbctx, role.ID)
+		if err != nil {
+			errs <- err
+			return
+		}
+	}()
+
+	wg.Wait()
+	close(errs)
+
+	for err := range errs {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.Role{}, err
+		}
+	}
+
+	resp := types.Role{
+		ID:      dbrole.ID,
+		Name:    dbrole.Name,
+		Actions: actions,
+
+		ResourceID: dbrole.ResourceID,
+		CreatedBy:  dbrole.CreatedBy,
+		UpdatedBy:  dbrole.UpdatedBy,
+		CreatedAt:  dbrole.CreatedAt,
+		UpdatedAt:  dbrole.UpdatedAt,
+	}
+
+	return resp, nil
+}
+
+func (e *engine) UpdateRoleV2(ctx context.Context, actor, roleResource types.Resource, newName string, newActions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.UpdateRoleV2")
+	defer span.End()
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	err = e.store.LockRoleForUpdate(dbCtx, roleResource.ID)
+	if err != nil {
+		sErr := fmt.Errorf("failed to lock role: %s: %w", roleResource.ID, err)
+
+		span.RecordError(sErr)
+		span.SetStatus(codes.Error, sErr.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	role, err := e.GetRoleV2(dbCtx, roleResource)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	newName = strings.TrimSpace(newName)
+
+	if newName == "" {
+		newName = role.Name
+	}
+
+	addActions, rmActions := diff(role.Actions, newActions)
+
+	// If no changes, return existing role
+	if newName == role.Name && len(addActions) == 0 && len(rmActions) == 0 {
+		if err = e.store.CommitContext(dbCtx); err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+		}
+
+		return role, nil
+	}
```

Comment:
If there's an error during commit, this will return a nil error.

Reply:
I actually thought about this, my intention was in the rare occasion this happens, since there was no changes, returning the values to the user rather than an error that has no effect would provide a better experience

Reply:
you're right. this is such a rare scenario I now think improving user experience in this very case is unnecessary.